// eslint-plugin-fp-ts/lib/rules/prefer-chain.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var experimental_utils_1 = require("@typescript-eslint/experimental-utils");
var fp_ts_1 = require("fp-ts");
var function_1 = require("fp-ts/function");
var utils_1 = require("../utils");
exports.default = utils_1.createRule({
    name: "prefer-chain",
    meta: {
        type: "suggestion",
        fixable: "code",
        hasSuggestions: true,
        schema: [],
        docs: {
            description: "Replace map + flatten with chain",
            recommended: "warn",
        },
        messages: {
            mapFlattenIsChain: "map followed by flatten can be replaced by chain",
            replaceMapFlattenWithChain: "replace map and flatten with chain",
        },
    },
    defaultOptions: [],
    create: function (context) {
        var isPipeOrFlowExpression = utils_1.contextUtils(context).isPipeOrFlowExpression;
        return {
            CallExpression: function (node) {
                function_1.pipe(node, isPipeOrFlowExpression, fp_ts_1.boolean.fold(function_1.constVoid, function () {
                    return function_1.pipe(utils_1.getAdjacentCombinators(node, {
                        name: /map|mapWithIndex/,
                        types: [experimental_utils_1.AST_NODE_TYPES.CallExpression],
                    }, {
                        name: "flatten",
                        types: [
                            experimental_utils_1.AST_NODE_TYPES.Identifier,
                            experimental_utils_1.AST_NODE_TYPES.MemberExpression,
                        ],
                    }, true), fp_ts_1.option.bindTo("combinators"), fp_ts_1.option.bind("mapCalleeIdentifier", function (_a) {
                        var mapNode = _a.combinators[0];
                        return utils_1.calleeIdentifier(mapNode);
                    }), fp_ts_1.option.map(function (_a) {
                        var _b = _a.combinators, mapNode = _b[0], flattenNode = _b[1], mapCalleeIdentifier = _a.mapCalleeIdentifier;
                        var chainIndentifier = mapCalleeIdentifier.name === "mapWithIndex"
                            ? "chainWithIndex"
                            : "chain";
                        context.report({
                            loc: {
                                start: mapNode.loc.start,
                                end: flattenNode.loc.end,
                            },
                            messageId: "mapFlattenIsChain",
                            suggest: [
                                {
                                    messageId: "replaceMapFlattenWithChain",
                                    fix: function (fixer) {
                                        return [
                                            fixer.remove(flattenNode),
                                            fixer.removeRange([
                                                mapNode.range[1],
                                                flattenNode.range[0],
                                            ]),
                                            fixer.replaceText(mapCalleeIdentifier, chainIndentifier),
                                        ];
                                    },
                                },
                            ],
                        });
                    }));
                }));
            },
        };
    },
});
